server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java

```diff
@@ -0,0 +1,384 @@
+        };
+
+        // firstly, validate core attributes like username
+        // TODO: we should be able to move most of the validation from legacy validators to this method
+        if (!validateCoreAttributes(attribute, listeners)) {
+            return false;
+        }
+
+        List<AttributeValidator> validators = this.validators.getOrDefault(attribute.getKey(), new ArrayList<>());
+
+        validators.addAll(this.validators.getOrDefault(READ_ONLY_ATTRIBUTE_KEY, Collections.emptyList()));
+
+        boolean success = true;
+
+        for (AttributeValidator validator : validators) {
+            if (!validator.getValidator().validate(attribute, user)) {
+                success = false;
+                if (listeners != null) {
+                    triggerValidationListeners(attribute, validator.getMessage(), listeners);
+                }
+            }
+        }
+
+        return success;
+    }
+
+    private boolean validateCoreAttributes(Entry<String, List<String>> attribute, BiConsumer<Entry<String, List<String>>, String>[] listeners) {
+        if (UserModel.USERNAME.equals(attribute.getKey()) && attribute.getValue() == EMPTY_VALUE) {
+            triggerValidationListeners(attribute, "missingUsernameMessage", listeners);
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean contains(String name) {
+        List<String> values = get(name);
+        return values != null && values != EMPTY_VALUE;
+    }
+
+    void setUser(UserModel user) {
+        this.user = user;
+    }
+
+    private void addValidator(AttributeValidator validator) {
+        validators.computeIfAbsent(validator.getAttributeName(), s -> new ArrayList<>()).add(validator);
+    }
+
+    private void configureValidators(UserProfileContext context,
+            Map<UserProfileContext, Function<KeycloakSession, List<AttributeValidator>>> validatorByContext, KeycloakSession session) {
+        // for now, build the validators whenever a new user profile is created. but we should be able to improve this can calculate
+        // only once. This part should change once we introduce the validation api.
+        Function<KeycloakSession, List<AttributeValidator>> validators = validatorByContext.get(context);
+
+        if (validators != null) {
+            for (AttributeValidator validator : validators.apply(session)) {
+                addValidator(validator);
+            }
+        }
+
+        if (UserProfileContext.USER_RESOURCE.equals(context)) {
+            addValidator(new AttributeValidator(READ_ONLY_ATTRIBUTE_KEY, UPDATE_READ_ONLY_ATTRIBUTES_REJECTED, isReadOnlyAttributeUnchanged(adminReadOnlyAttributesPattern)));
+        } else {
+            addValidator(new AttributeValidator(READ_ONLY_ATTRIBUTE_KEY, UPDATE_READ_ONLY_ATTRIBUTES_REJECTED, isReadOnlyAttributeUnchanged(readOnlyAttributesPattern)));
+        }
+    }
+
+    private Map<String, List<String>> transformAttributes(Map<String, ?> attributes) {
```

Comment:
Make this protected to allow customization please

Reply:
Another internal code and it changed to make more explicitly that we are normalizing the attributes for a profile.

At this regard, you should not need to override this method but rely on whatever configuration you have for a profile.